pkg/trace/obfuscate/json.go

```diff
@@ -25,27 +27,47 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
 	// as much of it as we could. It is safe to accept the output, even if partial.
 }
 
+type valueTransformer func(string) string
+
 type jsonObfuscator struct {
-	keepers map[string]bool // these keys will not be obfuscated
+	keepKeys      map[string]bool // these keys will not be obfuscated
+	transformKeys map[string]bool // these keys pass through the transformer
+	transformer   valueTransformer
 
 	scan     *scanner // scanner
 	closures []bool   // closure stack, true if object (e.g. {[{ => []bool{true, false, true})
 	key      bool     // true if scanning a key
 
-	wiped     bool // true if obfuscation string (`"?"`) was already written for current value
-	keeping   bool // true if not obfuscating
-	keepDepth int  // the depth at which we've stopped obfuscating
+	wiped             bool // true if obfuscation string (`"?"`) was already written for current value
+	keeping           bool // true if not obfuscating
+	transformingValue bool // true if collecting the next literal for transformation
+	keepDepth         int  // the depth at which we've stopped obfuscating
 }
 
-func newJSONObfuscator(cfg *config.JSONObfuscationConfig) *jsonObfuscator {
+func (o *Obfuscator) newJSONObfuscator(cfg *config.JSONObfuscationConfig) *jsonObfuscator {
 	keepValue := make(map[string]bool, len(cfg.KeepValues))
 	for _, v := range cfg.KeepValues {
 		keepValue[v] = true
 	}
+	var transformValues map[string]bool
+	var transformer valueTransformer
+	if cfg.TransformerType == "obfuscate_sql" {
+		transformer = o.ObfuscateSQLStringSafe
+	} else {
+		log.Warnf("unknown JSON value transformer type %s", cfg.TransformerType)
```

Comment:
Can we cleanly check that all values are correct and warn about them in `apply.go`? If that's complicated, it's fine to leave it here, but ideally we'd validate everything there.

Reply:
Simplified here https://github.com/DataDog/datadog-agent/commit/c5014db632ce98ccd97685e861f10007c58222dd